### PR TITLE
Headless/bsdhrubo/bc 109/comparison api

### PR DESCRIPTION
**Pull request by bsdhrubo**
# Description
(Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.)

## Type of change

(Please delete options that are not relevant.)

- [ ] Bug fix
- [x] New feature

# How Has This Been Tested?

(Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration)

- [x] Tested by myself
- [ ] Unit test

# Developer Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

# Review check list:

- [x] Separation of Concerns followed (Reviewer name/NaN)
- [x] Code is in sync with existing code patterns/technologies. (Reviewer name/NaN)
- [ ] Are there any obvious optimizations that will improve performance? (Reviewer name/NaN)
- [ ] Can any logging or debugging code be removed? (Reviewer name/NaN)
- [ ] Are all data inputs checked (for the correct type, length, format, and range) and encoded? (Reviewer name/NaN)
- [ ] Are invalid parameter values handled? (Reviewer name/NaN)
- [ ] Is error handling done the correct way? (Reviewer name/NaN)
- [ ] Are all the edge cases (null, 0, negative) handled properly? (Reviewer name/NaN)

### CODE REVIEW

File: packages/headless/src/modules/compare/services/index.ts [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    private helper: Helper,
+  ) {}
+
+  async addItemToCompare(
+    userId: string,
+    productId: string,
+  ): Promise<CompareResponse> {
+    if (!productId.trim()) {
+      return this.helper.serviceResponse.errorResponse(
+        'Invalid product id.',
```
Comment (Anjarul-Islam) [inline packages/headless/src/modules/compare/services/index.ts:20]: Can you please add these error message as enum in models and then use them in swagger and here

File: packages/models/src/compare/compare.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import { ErrorResponse, SuccessResponse } from "src/index";
+
+export interface AddCompareItem {
+    productId: string;
+}
+
+export interface CompareItems {
+    productId: string;
+    products?: any;
```
Comment (Anjarul-Islam) [inline packages/models/src/compare/compare.ts:9]: why any? don't we know the types of products?

Comment (bsdhrubo) [reply packages/models/src/compare/compare.ts:9]: I will define the type after adding product module.
It might be required to populate product in future.

File: packages/headless/src/modules/compare/rest/index.ts [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+@UseGuards(JwtAuthGuard)
+@ApiBearerAuth()
+@Controller('compare')
+export class CompareController {
+  constructor(private compareService: CompareService) {}
+
+  @ApiResponse({
+    description: 'Add product to compare',
+    type: CompareSuccessResponseDto,
+  })
```
Comment (ismail0946) [inline packages/headless/src/modules/compare/rest/index.ts:36]: Error responses should be shown in swagger.

File: packages/headless/src/database/mongodb/common/modelOptions.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+    timestamps: true,
+    versionKey: false,
+    toJSON: {
+      transform: (_, ret) => {
+        delete ret._id;
+        delete ret.createdAt;
+        delete ret.updatedAt;
+        return ret;
+      },
+    },
```
Comment (Anjarul-Islam) [inline packages/headless/src/database/mongodb/common/modelOptions.ts:12]: Seems good to use separate function to reduce code redundancy.
But for visibility we may skip this improvement. I've discussed with rezve vai. He also suggested to not overwrite toJSON options. and also he recommended to use the options in schema, rather than using from function response

Comment (bsdhrubo) [reply packages/headless/src/database/mongodb/common/modelOptions.ts:12]: Here only the toJSON and toObject options provides the transform method to overwrite the response doc.
Would you please elaborate,
If we need the same transformation for all of the schema what will be the problem in visibility? and
What will be the problem if we are overwrite toJSON options?

I didn't find any better solution by which we can transform the response doc without using the transform method. Ans this also requires less code. That's why I've implemented in this way. 
It would be very helpful if you can provide an example how we can use the options in schema.

File: packages/headless/config/database/index.ts [hunk 5]
```diff
@@ -1,17 +1,33 @@
...
-    },
-    db: DB || 'MONGO'
-}
+	mongodb: {
+		PROD_URI:
+			MONGODB_URI_PROD! || 'mongodb://localhost:27017/bs-commerce-prod',
+		DEV_URI:
+			MONGODB_URI_DEV! || 'mongodb://localhost:27017/bs-commerce-dev',
+		TEST_URI:
+			MONGODB_URI_TEST! || 'mongodb://localhost:27017/bs-commerce-test',
```
Comment (Anjarul-Islam) [inline packages/headless/config/database/index.ts:22]: Can't we use one URI? because our application will not run in different environment at the same time

File: packages/headless/config/core/index.ts [hunk 6]
```diff
@@ -2,7 +2,7 @@ const { PORT, API, ENV, HOST, REST_API_PREFIX } = process.env;
 export const coreConfig = {
     port: parseInt(PORT) || 3000,
     api: API || 'REST',
-    env: ENV || 'DEVELOPMENT',
+    env: ENV || 'DEV',
```
Comment (Anjarul-Islam) [inline packages/headless/config/core/index.ts:5]: wa can keep it as full text

File: packages/headless/jest-int.json [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+	"testEnvironment": "node",
+	"testRegex": ".spec.ts$",
+	"transform": {
+		"^.+\\.(t|j)s$": "ts-jest"
+	},
+	"moduleNameMapper": {
+		"src/(.*)": "<rootDir>/src/$1",
+		"config/(.*)$": "<rootDir>/config/$1"
+	}
+}
```
Comment (Anjarul-Islam) [inline packages/headless/jest-int.json:13]: if this file has no use case then we may delete this one

File: packages/headless/src/main.ts [hunk 8]
```diff
@@ -1,22 +1,21 @@
...
 async function bootstrap() {
-  await connectToDatabase(dbConfig.db as DB);
-  const app = await NestFactory.create(AppModule);
-  app.setGlobalPrefix(coreConfig.restApiPrefix);
-  app.useGlobalPipes(new ValidationPipe());
-  coreConfig.api === 'REST' ? SwaggerConfig(app) : null;
-  await app.listen(coreConfig.port);
-  console.log(`http://${coreConfig.host}:${coreConfig.port}`);
+	await connectToDatabase(dbConfig.db as DB);
+	const app = await NestFactory.create(AppModule);
```
Comment (ismail0946) [inline packages/headless/src/main.ts:13]: why did you change this file?

File: packages/headless/src/internal/graphql/graphql.init.ts [hunk 9]
```diff
@@ -1,14 +1,14 @@
...
-  return [
-    GraphQLModule.forRoot<ApolloDriverConfig>({
-      driver: ApolloDriver,
-      typePaths: ['src/**/*.graphql'],
-      path: '/graphql',
-      playground: (coreConfig.env === 'DEVELOPMENT') ? true : false,
-    }),
-  ];
+	return [
+		GraphQLModule.forRoot<ApolloDriverConfig>({
```
Comment (ismail0946) [inline packages/headless/src/internal/graphql/graphql.init.ts:7]: same as before

File: packages/headless/src/database/mongodb/connect.ts [hunk 10]
```diff
@@ -2,22 +2,22 @@ import { dbConfig } from 'config/database';
...
-    await Mongoose.connect(dbConfig.mongodb.URI!);
-    const { connection } = Mongoose;
+	await Mongoose.connect(dbConfig.mongodb.URI!);
+	const { connection } = Mongoose;
 
-    connection.on('connected', () => {
-        console.info('Success! Connected to MongoDB.');
-    });
+	connection.on('connected', () => {
+		console.info('Success! Connected to MongoDB.');
```
Comment (ismail0946) [inline packages/headless/src/database/mongodb/connect.ts:9]: why did you change this file?
Please don't change this file. revert it

File: packages/headless/src/database/database.init.ts [hunk 11]
```diff
@@ -1,23 +1,22 @@
...
-                throw new Error('No database found to connect');
-        }
-    }
-    catch (err) {
-        console.log(err)
-        console.error('Error connecting to database');
-    }
-}
\ No newline at end of file
+	try {
```
Comment (ismail0946) [inline packages/headless/src/database/database.init.ts:7]: why did you change this file?
Please don't change this file. revert it

File: packages/headless/config/core/index.ts [hunk 12]
```diff
@@ -1,8 +1,8 @@
 const { PORT, API, ENV, HOST, REST_API_PREFIX } = process.env;
 export const coreConfig = {
-    port: parseInt(PORT) || 3000,
-    api: API || 'REST',
-    env: ENV || 'DEVELOPMENT',
-    host: HOST || 'localhost',
-    restApiPrefix: REST_API_PREFIX || 'api'
-}
+	port: parseInt(PORT) || 3000,
```
Comment (ismail0946) [inline packages/headless/config/core/index.ts:3]: why did you change this file?
Please don't change this file. revert it

File: packages/headless/src/database/mongodb/compare/index.ts [hunk 13]
```diff
@@ -0,0 +1,62 @@
+import { Compare, CompareItems } from 'src/entity/compare';
+import { ICompareDatabase } from 'src/modules/compare/repositories/compare.db.interface';
+import { CompareModel } from './compare.model';
+
+export class CompareDatabase implements ICompareDatabase {
+	async getCompareListByUserId(userId: string): Promise<Compare | null> {
+		const compareList = await CompareModel.findOne({
```
Comment (ismail0946) [inline packages/headless/src/database/mongodb/compare/index.ts:7]: you can return just the first line.

File: packages/headless/src/modules/compare/rest/index.ts [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+		description: 'Add product to compare Error Response',
+		type: CompareErrorResponseDto,
+		status: HttpStatus.BAD_REQUEST,
+	})
+	@ApiResponse({
+		description: 'Delete compare by compare id',
+		type: CompareSuccessResponseDto,
+		status: HttpStatus.OK,
+	})
+	@ApiQuery({ name: 'compareId', example: '' })
```
Comment (ismail0946) [inline packages/headless/src/modules/compare/rest/index.ts:105]: if you don't pass any example value then you may remove this example property.

File: packages/headless/src/modules/compare/services/test.service.ts [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+        if (data) {
+            const result = successResponse(CompareDataDto, data);
+            return result;
+        } else {
+            return errorResponse(GetCompareErrorEnum.COMPARISON_LIST_IS_EMPTY);
+        }
+    }
+
+    async getCompare(userId: string): Promise<CompareResponse<boolean>> {
+        const data = (await this.compareRepository.getCompareByUserId(userId)) as CompareDataDto;
```
Comment (ismail0946) [inline packages/headless/src/modules/compare/services/test.service.ts:21]: as CompareDataDto
why is this used?

File: packages/headless/src/modules/compare/services/test.service.ts [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+        if (data) {
+            // const result = successResponse(true);
```
Comment (ismail0946) [inline packages/headless/src/modules/compare/services/test.service.ts:23]: remove unnecessary comment

File: packages/headless/test/modules/auth/auth.service.spec.ts [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+import {
+  connectTestDatabase,
+  signInDemoUser,
+  signUpDemoUser,
+  testProductId,
+  testTimeout,
+  testUserId,
+} from '../../test-utility';
+
+// const token = getDemoUserToken(testUserId, testUsername).token;
```
Comment (ismail0946) [inline packages/headless/test/modules/auth/auth.service.spec.ts:15]: same

File: packages/headless/test/modules/auth/auth.service.spec.ts [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+  //   describe('Creating a new user....', () => {
+  //     it('should return user object', async () => {
+  //       const response = await service.signUp(signUpUser);
+  //       expect(response.code).toEqual(201);
+  //     });
+  //   });
+  describe('Signing in user....', () => {
+    it('should return token object', async () => {
+      const response = await service.signIn(signInUser);
+      //   console.log({ response });
```
Comment (ismail0946) [inline packages/headless/test/modules/auth/auth.service.spec.ts:52]: remove console.log

File: packages/headless/test/test-utility/test.databse.ts [hunk 19]
```diff
@@ -0,0 +1,27 @@
+import * as Mongoose from 'mongoose';
+
+export const connectTestDatabase = async (): Promise<void> => {
+	const mongoTestUri = 'mongodb://localhost:27017/bs-commerce-test';
```
Comment (ismail0946) [inline packages/headless/test/test-utility/test.databse.ts:4]: you should use this URI from the .env file.

File: packages/models/src/compare/compare.ts (hunk 2 above)
Comment (ismail0946) [reply packages/models/src/compare/compare.ts:14]: now you can integrate with the product module.
